### PR TITLE
Assume: Use plain FIRRTL assume + format string instead of printf + assume.

### PR DESCRIPTION
**Pull request by dtzSiFive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Backend Code Generation

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

chisel3.Assume now directly emits a FIRRTL `assume` with the specified format message.
Previously a `when not(predicate): printf(...)` was emitted followed by a message-less `assume`.
The resulting SystemVerilog will change, more aligned with `assert` output but specifically uses `$error(..)` instead of a printf guarded by the usual printf guards.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

-----

Example output before/after:

Before:

```
FIRRTL version 4.0.0
circuit SimpleTest :
  public module SimpleTest : @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    input clock : Clock @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    input reset : UInt<1> @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    output io : { flip in : UInt<8>, out : UInt<8>} @[src/test/scala/chiselTests/VerificationSpec.scala 12:14]

    connect io.out, io.in @[src/test/scala/chiselTests/VerificationSpec.scala 16:10]
    node _T = eq(io.in, UInt<2>(0h3)) @[src/test/scala/chiselTests/VerificationSpec.scala 18:14]
    when _T : @[src/test/scala/chiselTests/VerificationSpec.scala 18:23]
      node _T_1 = neq(io.in, UInt<2>(0h2)) @[src/test/scala/chiselTests/VerificationSpec.scala 19:18]
      node _T_2 = eq(reset, UInt<1>(0h0)) @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
      when _T_2 : @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
        node _T_3 = eq(_T_1, UInt<1>(0h0)) @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
        when _T_3 : @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
          printf(clock, UInt<1>(0h1), "Assumption failed\n    at VerificationSpec.scala:19 assume(io.in =/= 2.U)\n") : printf @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
        assume(clock, _T_1, UInt<1>(0h1), "") : assume @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
```
```systemverilog
// Generated by CIRCT firtool-1.66.0-55-gcad1b4fd1
// Standard header to adapt well known macros for prints and assertions.

// Users can define 'PRINTF_COND' to add an extra gate to prints.
`ifndef PRINTF_COND_
  `ifdef PRINTF_COND
    `define PRINTF_COND_ (`PRINTF_COND)
  `else  // PRINTF_COND
    `define PRINTF_COND_ 1
  `endif // PRINTF_COND
`endif // not def PRINTF_COND_

module SimpleTest(	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
  input        clock,	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
               reset,	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
  input  [7:0] io_in,	// src/test/scala/chiselTests/VerificationSpec.scala:12:14
  output [7:0] io_out	// src/test/scala/chiselTests/VerificationSpec.scala:12:14
);

  wire _GEN = io_in != 8'h2;	// src/test/scala/chiselTests/VerificationSpec.scala:19:18
  wire _GEN_0 = io_in == 8'h3 & ~reset;	// src/test/scala/chiselTests/VerificationSpec.scala:18:14, :19:11
  `ifndef SYNTHESIS	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
    always @(posedge clock) begin	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
      if ((`PRINTF_COND_) & _GEN_0 & ~_GEN)	// src/test/scala/chiselTests/VerificationSpec.scala:19:{11,18}
        $fwrite(32'h80000002,
                "Assumption failed\n    at VerificationSpec.scala:19 assume(io.in =/= 2.U)\n");	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
    end // always @(posedge)
  `endif // not def SYNTHESIS
  always @(posedge clock) begin	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
    if (_GEN_0)	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
      assume__assume: assume(_GEN);	// src/test/scala/chiselTests/VerificationSpec.scala:19:{11,18}
  end // always @(posedge)
  assign io_out = io_in;	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
endmodule
```

After:

```
FIRRTL version 4.0.0
circuit SimpleTest :
  public module SimpleTest : @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    input clock : Clock @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    input reset : UInt<1> @[src/test/scala/chiselTests/VerificationSpec.scala 11:7]
    output io : { flip in : UInt<8>, out : UInt<8>} @[src/test/scala/chiselTests/VerificationSpec.scala 12:14]

    connect io.out, io.in @[src/test/scala/chiselTests/VerificationSpec.scala 16:10]
    node _T = eq(io.in, UInt<2>(0h3)) @[src/test/scala/chiselTests/VerificationSpec.scala 18:14]
    when _T : @[src/test/scala/chiselTests/VerificationSpec.scala 18:23]
      node _T_1 = neq(io.in, UInt<2>(0h2)) @[src/test/scala/chiselTests/VerificationSpec.scala 19:18]
      node _T_2 = eq(reset, UInt<1>(0h0)) @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
      when _T_2 : @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
        assume(clock, _T_1, UInt<1>(0h1), "Assumption failed\n    at VerificationSpec.scala:19 assume(io.in =/= 2.U)\n") : assume @[src/test/scala/chiselTests/VerificationSpec.scala 19:11]
```
```systemverilog
// Generated by CIRCT firtool-1.66.0-55-gcad1b4fd1
module SimpleTest(	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
  input        clock,	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
               reset,	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
  input  [7:0] io_in,	// src/test/scala/chiselTests/VerificationSpec.scala:12:14
  output [7:0] io_out	// src/test/scala/chiselTests/VerificationSpec.scala:12:14
);

  always @(posedge clock) begin	// src/test/scala/chiselTests/VerificationSpec.scala:19:11
    if (io_in == 8'h3 & ~reset)	// src/test/scala/chiselTests/VerificationSpec.scala:18:14, :19:11
      assume__assume:
        assume(io_in != 8'h2)
        else $error("Assumption failed\n    at VerificationSpec.scala:19 assume(io.in =/= 2.U)\n");	// src/test/scala/chiselTests/VerificationSpec.scala:19:{11,18}
  end // always @(posedge)
  assign io_out = io_in;	// src/test/scala/chiselTests/VerificationSpec.scala:11:7
endmodule
```

cc @fabianschuiki @uenoku

### CODE REVIEW

**Comment by jackkoenig** [comment]
I would say this is more "Backend Code Generation" than "API Modification". Still orange and important to note 🙂

**Comment by dtzSiFive** [comment]
> I would say this is more "Backend Code Generation" than "API Modification". Still orange and important to note 🙂

I'll update this, thank you, wasn't sure :+1: .  Makes sense to me!

File: core/src/main/scala/chisel3/VerificationStatement.scala [hunk 1]
```diff
@@ -432,23 +432,33 @@ private object VerificationStatement {
...
-  ): Unit = {
+  ): Printable = {
     val (filename, line, content) = lineInfo
     val lineMsg = s"$filename:$line $content".replaceAll("%", "%%")
-    val fmt = message match {
+    message match {
       case Some(msg) =>
         p"$kind failed: $msg\n    at $lineMsg\n"
       case None => p"$kind failed\n    at $lineMsg\n"
     }
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/VerificationStatement.scala:449]: Incidentally I'm about to PR getting rid of lineMsg, let's get your PR in first.

**Review by jackkoenig — APPROVED** [review]
LGTM, nice improvement!

**Review by fabianschuiki — APPROVED** [review]
Very nice improvement!

File: core/src/main/scala/chisel3/VerificationStatement.scala (hunk 1 above)
Comment (dtzSiFive) [reply core/src/main/scala/chisel3/VerificationStatement.scala:449]: Great, and thanks for letting this go first, sorry for the delay :+1: .

**Pull request merged**